core/embed/trezorhal/stm32f4/usb/usbd_internal.h

```diff
@@ -0,0 +1,138 @@
+  uint8_t bMaxPacketSize0;
+  uint16_t idVendor;
+  uint16_t idProduct;
+  uint16_t bcdDevice;
+  uint8_t iManufacturer;
+  uint8_t iProduct;
+  uint8_t iSerialNumber;
+  uint8_t bNumConfigurations;
+} usb_device_descriptor_t;
+
+typedef struct __attribute__((packed)) {
+  uint8_t bLength;
+  uint8_t bDescriptorType;
+  uint16_t wData;
+} usb_langid_descriptor_t;
+
+typedef enum {
+  USB_LANGID_ENGLISH_US = 0x409,
+} usb_language_id_t;
+
+typedef struct __attribute__((packed)) {
+  uint8_t bLength;
+  uint8_t bDescriptorType;
+  uint16_t wTotalLength;
+  uint8_t bNumInterfaces;
+  uint8_t bConfigurationValue;
+  uint8_t iConfiguration;
+  uint8_t bmAttributes;
+  uint8_t bMaxPower;
+} usb_config_descriptor_t;
+
+typedef struct __attribute__((packed)) {
+  uint8_t bLength;
+  uint8_t bDescriptorType;
+  uint8_t bInterfaceNumber;
+  uint8_t bAlternateSetting;
+  uint8_t bNumEndpoints;
+  uint8_t bInterfaceClass;
+  uint8_t bInterfaceSubClass;
+  uint8_t bInterfaceProtocol;
+  uint8_t iInterface;
+} usb_interface_descriptor_t;
+
+typedef struct __attribute__((packed)) {
+  uint8_t bLength;
+  uint8_t bDescriptorType;
+  uint8_t bFirstInterface;
+  uint8_t bInterfaceCount;
+  uint8_t bFunctionClass;
+  uint8_t bFunctionSubClass;
+  uint8_t bFunctionProtocol;
+  uint8_t iFunction;
+} usb_interface_assoc_descriptor_t;
+
+typedef struct __attribute__((packed)) {
+  uint8_t bLength;
+  uint8_t bDescriptorType;
+  uint8_t bEndpointAddress;
+  uint8_t bmAttributes;
+  uint16_t wMaxPacketSize;
+  uint8_t bInterval;
+} usb_endpoint_descriptor_t;
+
+// Number of reserved bytes for the state of each class.
+#define USBD_CLASS_STATE_MAX_SIZE 128
+
+// Returns the pointer to class state structure reserved for the
+// specified interface number.
+//
+// The function checks if the interface number is valid and the type
+// matches and returns NULL if not. If the `class` is NULL, the function
+// returns the valid pointer only if the slot is empty.
+//
+// The returned array has `USBD_CLASS_STATE_MAX_SIZE` bytes
+// and is aligned to 8-byte boundary.
+void *usb_get_iface_state(uint8_t iface_num, const USBD_ClassTypeDef *class);
+
+// Assigns the concrete class to the slot `iface_num`.
+// If the class is NULL the slot is cleared.
+void usb_set_iface_class(uint8_t iface_num, const USBD_ClassTypeDef *class);
+
+// Returns the pointer to the next free bytes in the buffer
+// reserved for the class driver descriptors (interface, endpoint, ...)
+//
+// The function checks if the remaining space is enough and
+// returns NULL if not.
+void *usb_desc_alloc_iface(size_t desc_len);
+
+// Adds the class driver descriptors to the the configuration descriptor and
+// updates configuration descriptor length and number of interfaces.
+//
+// The class driver must call this function after all the its
+// descriptors have been written to the memory preallocated
+// by `usb_desc_alloc_iface`.
+void usb_desc_add_iface(size_t desc_len, uint8_t num_ifaces);
```

Comment:
adding the possibility to add more interfaces would warrant name change to in my usb_desc_add_ifaces opinion. And possibly rename the desc_len to total_desc_len?

Reply:
Finally, I decided to completely redo this part and get rid of this function altogether, see 12c155f6c1a9dda053535c8bbfbb6d6161fb5ace